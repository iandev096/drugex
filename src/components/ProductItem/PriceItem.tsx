import { formatDistance, formatRelative } from "date-fns";
import { View } from "react-native";
import { Caption, List, Text } from "react-native-paper";

type Props = {
  price: number;
  date: string;
}

export default function PriceItem({ price, date }: Props) {

  return (
    <List.Item
      title={`$${price}`}
      right={() => <View>
        <Text>{formatRelative(new Date(date), new Date())}</Text>
        <Caption>{formatDistance(new Date(date), new Date())}</Caption>
      </View>}
      testID='listItem'
    />
  )
}