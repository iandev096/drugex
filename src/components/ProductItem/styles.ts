import { StyleSheet } from "react-native";
import { Theme } from "../../../theme";

const styles = (theme: Theme) =>
  StyleSheet.create({
    productItem: {
      position: "relative"
    },
    buttonGroup: {
      position: "absolute",
      right: 10,
      bottom: 0,
      top: 0,
      flexDirection: "row",
      justifyContent: "flex-end",
      alignItems: "center",
    },
    border: {
      borderBottomWidth: 3,
      borderColor: theme.colors.border,
    },
    listAccordion: { paddingVertical: 16 },
    icon: { paddingHorizontal: 8 },
    pastPrices: { fontSize: 12, textTransform: "uppercase", fontWeight: '600' }
  });

export default styles;
