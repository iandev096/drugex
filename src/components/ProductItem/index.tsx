import { formatDistance, formatRelative } from "date-fns";
import React from "react";
import { View } from "react-native";
import { IconButton, List } from "react-native-paper";
import { Price } from "../../contexts/app/type";
import useAppTheme from "../../hooks/useAppTheme";
import styles from "./styles";

export type ProductItemProps = {
  name: string;
  prices: Price[];
  expanded: boolean;
  last?: boolean;
  onPress?: () => void;
  onEdit?: () => void;
  onDelete?: () => void;
};

export default function ProductItem({
  name,
  expanded,
  prices,
  last,
  onPress,
  onDelete,
  onEdit,
}: ProductItemProps) {
  const theme = useAppTheme();

  return (
    <View style={[styles(theme).productItem, !last && styles(theme).border]}>
      <List.Accordion
        right={!expanded ? () => <></> : undefined}
        title={`${name} -  $${prices[prices.length - 1]?.price}`}
        left={(props) => <List.Icon {...props} icon="apps" />}
        expanded={expanded}
        onPress={onPress}
        testID='btn:onPress'
      >
        <List.Item
          title={
            prices.length === 1 ? "No Past Price Available" : "Past Price(s)"
          }
          titleStyle={{ fontSize: 12, textTransform: "uppercase" }}
        />
        {prices
          .map((priceItem) => (
            <List.Item
              title={`$${priceItem.price}`}
              description={`${formatRelative(
                new Date(priceItem.date),
                new Date()
              )} - ${formatDistance(new Date(priceItem.date), new Date())} `}
              key={priceItem.id}
              testID='listItem'
            />
          ))
          .reverse()
          .slice(1)}
      </List.Accordion>
      {!expanded && (
        <View style={styles(theme).buttonGroup}>
          <IconButton  icon="lead-pencil" onPress={onEdit} testID='btn:onEdit' />
          <IconButton icon="trash-can" onPress={onDelete} testID='btn:onDelete' />
        </View>
      )}
    </View>
  );
}
