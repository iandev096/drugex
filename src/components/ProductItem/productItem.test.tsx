import { fireEvent, render } from '@testing-library/react-native'
import ProductItem, { ProductItemProps } from '.';

const props: ProductItemProps = {
    expanded: false,
    name: 'Paracetamol',
    prices: [...new Array(4)]
        .map((_, idx) => ({ id: idx, date: new Date().toISOString(), price: 20 })),
    last: false,
    onDelete: () => console.log('onDelete triggered'),
    onEdit: () => console.log('onEdit triggered'),
    onPress: () => console.log('onPress triggered')
};

test('displays correct product name', () => {
    const { getByText } = render(<ProductItem {...props} />)

    expect(getByText(new RegExp(props.name, 'i'))).toBeTruthy();
})


test('displays correct product price', () => {
    const { getByText } = render(<ProductItem {...props} />)
    
    const latestPriceRegex = new RegExp(props.prices[props.prices.length - 1].price.toString(), 'i');
   
    expect(getByText(latestPriceRegex)).toBeTruthy();
})


test('all buttons work', () => {
    const { getByTestId } = render(<ProductItem {...props} />)

    const [listAccord, editBtn, deleteBtn] = [getByTestId('btn:onPress'), getByTestId('btn:onEdit'), getByTestId('btn:onDelete')];
    const logSpy = jest.spyOn(console, 'log');
    fireEvent.press(listAccord);
    fireEvent.press(editBtn);
    fireEvent.press(deleteBtn);

    expect(logSpy).toHaveBeenCalledTimes(3);
});

test('renders correctly', () => {
    const { toJSON } = render(<ProductItem {...props} />)

    expect(toJSON()).toMatchSnapshot();
});

test('displays a list of past prices when expand is true', () => {
    const { getAllByTestId } = render(<ProductItem {...props} expanded={true} />)

    const listItems = getAllByTestId('listItem');

    expect(listItems).toHaveLength(props.prices.length - 1);
});