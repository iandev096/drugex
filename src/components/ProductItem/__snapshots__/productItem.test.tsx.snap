// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "position": "relative",
      },
      Object {
        "borderBottomWidth": 3,
        "borderColor": undefined,
      },
    ]
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#f6f6f6",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
            "expanded": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "padding": 8,
              },
              Object {
                "paddingVertical": 16,
              },
            ],
          ]
        }
        testID="btn:onPress"
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "paddingHorizontal": 8,
              }
            }
          >
            <RNSVGSvgView
              bbHeight={24}
              bbWidth={24}
              fill="none"
              focusable={false}
              height={24}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
              width={24}
            >
              <RNSVGGroup
                fill={null}
                propList={
                  Array [
                    "fill",
                  ]
                }
              >
                <RNSVGPath
                  d="M21 0H3a3 3 0 0 0-3 3v18a3 3 0 0 0 3 3h18a3 3 0 0 0 3-3V3a3 3 0 0 0-3-3Z"
                  fill={4294755072}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
                <RNSVGPath
                  d="M5.006 19.032a3.486 3.486 0 0 1 0-4.93L9.508 9.6l4.93 4.93-4.502 4.502a3.486 3.486 0 0 1-4.93 0Z"
                  fill={4294922496}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
                <RNSVGPath
                  d="M18.953 5.006a3.486 3.486 0 0 0-4.93 0L9.46 9.571l4.93 4.93 4.564-4.565a3.486 3.486 0 0 0 0-4.93Z"
                  fill={4294967295}
                  opacity={0.9}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
                <RNSVGPath
                  d="m10.719 8.31-1.26 1.26 4.93 4.93 1.26-1.26-4.93-4.93Z"
                  fill={4294967295}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
                <RNSVGPath
                  d="m11.142 7.906-1.26 1.26 4.93 4.93 1.26-1.26-4.93-4.93Z"
                  fill={4278190080}
                  opacity={0.1}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
                <RNSVGPath
                  d="m10.719 8.33-1.26 1.26 4.93 4.93 1.26-1.26-4.93-4.93Z"
                  fill={4294967295}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Paracetamol -  $20
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "position": "absolute",
        "right": 10,
        "top": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 16.5,
              "height": 33,
              "width": 33,
            },
            undefined,
            undefined,
          ],
        ]
      }
      testID="btn:onEdit"
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 22,
              },
              Array [
                Object {
                  "lineHeight": 22,
                  "transform": Array [
                    Object {
                      "scaleX": 1,
                    },
                  ],
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ],
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          Û∞ôè
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 16.5,
              "height": 33,
              "width": 33,
            },
            undefined,
            undefined,
          ],
        ]
      }
      testID="btn:onDelete"
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 22,
              },
              Array [
                Object {
                  "lineHeight": 22,
                  "transform": Array [
                    Object {
                      "scaleX": 1,
                    },
                  ],
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ],
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          Û∞©π
        </Text>
      </View>
    </View>
  </View>
</View>
`;
